document/js-api/index.bs

```diff
@@ -0,0 +1,840 @@
+        1. Let |v| be ? [=Get=](|o|, |componentName|)
+        1. If |externtype| is of the form 𝖿𝗎𝗇𝖼 |functype|,
+            1. If [=IsCallable=](|v|) is false, throw a {{LinkError}} exception.
+            1. If |v| has a \[[FunctionAddress]] internal slot, and therefore is an [=Exported Function=],
+                1. Let |funcaddr| be the value of |v|'s \[[FunctionAddress]] internal slot.
+
+                Note: The signature is checked by [=instantiate_module=] invoked below.
+
+                2. [=Append=] |v| to |funcs|.
+            1. Otherwise,
+                1. [=Create a host function=] from |v| and let |funcaddr| be the result.
+            1. Let |externfunc| be the [=external value=] 𝖿𝗎𝗇𝖼 |funcaddr|
+            1. [=Append=] |externfunc| to |imports|.
+        1. If |externtype| is of the form 𝗀𝗅𝗈𝖻𝖺𝗅 |globaltype|,
+            1. If |globaltype| is [=i64=] or [=Type=](|v|) is not [=Number=], throw a {{LinkError}} exception.
+            1. Let |value| be [=ToWebAssemblyValue=](|v|, |globaltype|.valtype)
+            1. Assert: |globaltype|.mut is const, as verified by WebAssembly validation.
+            1. Let |store| be the current agent cluster's [=associated store=].
+            1. Let (|store|, |globaladdr|) be [=alloc_global=](|store|, |globaltype|, |value|).
+            1. Set the current agent cluster's [=associated store=] to |store|.
+            1. Let |externglobal| be 𝗀𝗅𝗈𝖻𝖺𝗅 |globaladdr|.
+            1. [=Append=] |externglobal| to |imports|.
+        1. If |externtype| is of the form 𝗆𝖾𝗆  |memtype|,
+            1. If |v| is not a {{Memory}} object, throw a {{LinkError}} exception.
+
+            Note: [=instantiate_module=] invoked below will check the imported {{Memory}}'s size against the importing module's requirements.
+
+            2. [=Append=] |v| to |memories|.
+            1. Let |externmem| be the [=external value=] 𝗆𝖾𝗆  |v|.\[[Memory]].
+            1. [=Append=] |externmem| to |imports|.
+        1. Otherwise, |externtype| is of the form 𝗍𝖺𝖻𝗅𝖾 |tabletype|,
+            1. If |v| is not a {{Table}} instance, throw a {{LinkError}} exception.
+
+            Note: The table's length, etc. is checked by [=instantiate_module=] invoked below.
+
+            2. [=Append=] |v| to |tables|.
+            1. Let |tableaddr| be |v|.\[[Table]]
+            1. Let |externtable| be the [=external value=] 𝗍𝖺𝖻𝗅𝖾 |tableaddr|.
+            1. [=Append=] |externtable| to |imports|.
+            1. For each element |func| of |v|.\[[Values]],
+                1. If |func| is not null, append |func| to |funcs|.
+    1. Let (|store|, |instance|) be [=instantiate_module=](|store|, |module|, |imports|), up through step 16 of [=the instantiation algorithm=], but omitting execution of the start function.
+    1. If |instance| is [=error=], throw a {{LinkError}} exception, or other <a href="#errors">appropriate exception type</a>.
+    1. For each |tableaddr| in |instance|.𝗍𝖺𝖻𝗅𝖾𝖺𝖽𝖽𝗋𝗌,
```

Comment:
This text is all about caching exported functions and tables so we never create multiple copies of them ever; all tables, not just exported tables, are iterated over to reduce duplication in this process. I think I've convinced myself that there's no way to create multiple JS objects in the same agent that represent the things. If a Wasm module somehow imported a table and exported a function from the table, it would create multiple copies. I'm not sure if this is possible with current Wasm instructions, but I can only imagine @rossberg's earlier text was written with this scenario in mind.

I don't really like how much this text depends on invariants of what's possible today. For example, if Wasm modules could import each other directly, that would give a different way of sharing functions and we might need to revisit this text. The simplification you are suggesting would be another way to get at the same thing.

I was thinking of rewriting this spec text to make a mapping from funcaddrs, tableaddrs, etc to JavaScript objects. If we specify a way in the future that these addresses may be GC'd, that specification would delete elements from the mapping. Then, the spec text for constructing these various objects would consult the mapping to find the appropriate object, or create a new one and cache it. I believe this would achieve the same observable effect as the current spec text. What do you think of it?

Reply:
Yeah, that simplification sounds great.

Reply:
Will implement in a soon-to-come follow-on patch.

Reply:
I'd be happy with that rewrite. I think the current formulation was done that way because we (I?) weren't sure that we'd actually require global caching, but given how the mapping is typically implemented it seems fine.